src/bap_main.ml

```diff
@@ -95,27 +95,48 @@ let process options project =
       | `file dst,fmt,ver ->
         Out_channel.with_file dst ~f:(fun ch ->
             Project.Io.save ~fmt ?ver ch project)
-      | `stdout,fmt,ver -> Project.Io.show ~fmt ?ver project)
+      | `stdout,fmt,ver ->
+        Project.Io.show ~fmt ?ver project)
+
+let parse_filename filename =
+  let fmt = match String.index filename '.' with
+    | None -> filename
+    | Some n -> String.subo ~pos:(n+1) filename in
+  match Bap_fmt_spec.parse fmt with
+  | `Error _ -> None, None
+  | `Ok (_,fmt,ver) -> Some fmt, ver
 
 let main o =
-  let digest = digest o in
-  let project = match Project.Cache.load digest with
+  let proj_of_input input =
+    let rooter = rooter o
+    and brancher = brancher o
+    and reconstructor = reconstructor o
+    and symbolizer = symbolizer o in
+    Project.create input ~disassembler:o.disassembler
+      ?brancher ?rooter ?symbolizer ?reconstructor  |> function
+    | Error err -> raise (Failed_to_create_project err)
+    | Ok project ->
+      Project.Cache.save (digest o) project;
+      project in
+  let proj_of_file ?ver ?fmt file =
+    In_channel.with_file file
+      ~f:(fun ch -> Project.Io.load ?fmt ?ver ch) in
+  let project = match Project.Cache.load (digest o) with
     | Some proj ->
       Project.restore_state proj;
       proj
-    | None ->
-      let rooter = rooter o
-      and brancher = brancher o
-      and reconstructor = reconstructor o
-      and symbolizer = symbolizer o in
-      let input =
+    | None -> match o.source with
+      | `File "builtin" ->
+        let fmt,ver = parse_filename o.filename in
+        proj_of_file ?fmt ?ver o.filename
+      | `File fmt ->
+        proj_of_file ~fmt o.filename
```

Comment:
Ok, we have lots of confusion here. Originally,  `File <name>` mapped to the `<name>-custom` where `<name>` "should be a name of factory method registered for a file source in project factory", as per `bap --help` documentation. It is not a format name, it is a name of the method, i.e., the loader name. 

Your interpretation basically makes the custom syntax just another way to specify a format, so we have two equivalent variants:
```
bap ./exe.marshal --source-type=project
bap ./exe --source-type=marshal-custom
```

It looks like that we don't need the `*-custom` form anymore, as it is superseded by the `loader` option. 

So, please, can you rectify this by:

1. removing the custom form from the source-type parse
2. fixing the ouptut representation, by changing `File of string` to `Project`
3. updating the source-type documentation